domain-knowledge/DOMAIN-OBJECTS.md

```diff
@@ -0,0 +1,501 @@
+- FLAC encoding and decoding tools are cost-free and open-source software.
+- FLAC is technically superior to most, if not all, competing formats, offering
+ the most features and flexibility.
+ 
+When customers purchase music from the catalog, they are able to download their
+purchased music in FLAC format, in addition to several other formats:
+
+- WAV
+- MP3
+- AAC
+
+#### Relationships
+
+- `FlacFile` `hasMany()` `Song`
+
+#### Access Policy
+
+- Any `User` that is associated with a `CatalogEntity` whose `approved_at` field
+  is non-null, and whose `deleted_at` field is null, may create a `FlacFile`.
+- Anybody may read a `FlacFile` record from the _database_, but not necessarily
+  the corresponding binary file from the filesystem.
+- Only customers who have purchased access to a given `FlacFile` may retrieve
+  the associated binary data from the filesystem.
+- The `User` who is associated with a given `CatalogEntity` may delete a `FlacFile`
+  that is under its control, but only under the following conditions:
+  - The `FlacFile` is being overwritten and is not associated with any other `Song`.
+    - If the `FlacFile` _is_ associated with another `Song`, the `User` is
+      prompted: "The audio file that you're replacing is associated with other
+      songs. Which of those songs, if any, should be updated with this audio file?"
+      (The `User` is presented with a list, and appropriate UI controls, to
+      select which `Songs'` relationships, if any, should be updated.) If the
+      `User` selects every `Song`, the "orphaned" `FlacFile` is deleted, along
+      with its binary filesystem data.
+  - A `Song` is deleted that would cause the `FlacFile` to be "orphaned"; in such
+    cases, the `FlacFile` is deleted, along with the associated binary filesystem
+    data.
+
+### Genre
+
+One or more `Genres` may be assigned to an `Album`. Assigning one or
+more `Genres` to an `Album` enables the `Album` to appear in search results for
+the associated `Genre`, thereby making it easier for potential customers to find
+music in their preferred `Genre`.
+
+The service provides a pre-set list of `Genres` derived from the ID3v1 MP3 tag
+specification. Additionally, `Artists` may request that new `Genres` are added,
+which, upon request, are queued for administrative approval. This policy prevents
+`Artists` from adding duplicate `Genres` (i.e., with alternate spellings) and
+from using gratuitous profanity, inconsistent formatting, etc.
+
+#### Relationships
+
+- `Genre` `belongsToMany()` `Album`
+
+#### Access Policy
+
+- Anybody may read `Genres` whose `approved_at` field is non-null.
+- Only service operator staff members may read `Genres` whose `approved_at` field
+  is null.
+- Any `CatalogEntity` may create a new `Genre`, _however_, a service operator
+  staff member must approve the record before it is visible to anybody else.
+  Upon approval, the `approved_at` field is set to the current date, and the
```

Comment:
What happens if it is disapproved?

Do you need to have a backup or will it be nullified?

Reply:
Another great question/point.

At a minimum, it seems prudent to notify the requester that the new `Genre` was denied, and for what reason (possible reasons would include only a virtual duplicate [variation only in spacing, spelling, etc.], profanity, malicious nonsense, etc.).

Given that the `Artist` (or `Label`, provided we add `Genres` to `Labels`) is able to edit this metadata at any time, they can simply choose an alternate in such a scenario.

Reply:
But is Genre required?

I assume they create a song select a genre and if it is a new one it is optimistically selected for that song but if it is declined then what?
Or do they have to apply first and then they can select for the song later? Seems a bit bad UX. But maybe it is necessary this way.

Reply:
I see your point, now.

No, the `Genre` is not required, and, yes, the new one is optimistically selected for that `Album`. (Just for the sake of clarity, `Genre` is assigned at the `Album` [and potentially `Label`] level, not `Song`.)

We could certainly allow the user to select a "fallback" `Genre` that is used if the proposed one is rejected, but is this more complicated than simply removing the rejected `Genre` from the `Album` and notifying the user of the rejection, and suggesting that they edit the `Album` choose a new one?

I'm fine with either, but if we support the fallback capability, the data structures and underlying logic become more complicated, so we'll have to work through that.

With the focus on an MVP, should we table this, for now, and run with what we have?

Reply:
Actually, I had forgotten that we changed this relationship to allow for more than one `Genre` per `Album`. So, it would be easy and harmless to allow (and perhaps encourage) the user to choose one or more additional `Genres` in case the proposed `Genre` is rejected.

Maybe all that's necessary here is some verbiage in the UI that explains the above concisely.